bpf/bpf_netdev.c

```diff
@@ -110,22 +111,75 @@ reverse_proxy6(struct __sk_buff *skb, int l4_off, struct ipv6hdr *ip6, __u8 nh)
 			return DROP_CSUM_L4;
 	}
 
+	/* Packets which have been translated back from the proxy must
+	 * skip any potential ingress proxy at the endpoint
+	 */
+	skb->tc_index |= TC_INDEX_F_SKIP_PROXY;
+
 	return 0;
 }
+#endif
+
+#ifdef FROM_HOST
+static inline void __inline__ handle_identity_from_proxy(struct __sk_buff *skb, __u32 *identity)
+{
+	int ret;
+
+	/* For packets from the proxy the identity can be specified via
+	 * skb->mark */
+	if ((ret = mark_is_from_proxy(skb))) {
+		*identity = get_identity_via_proxy(skb);
+
+		/* Packets from the ingress proxy must skip the proxy when the
+		 * destination endpoint evaluates the policy. As the packet
+		 * would loop otherwise. */
+		if (ret == SOURCE_INGRESS_PROXY)
+			skb->tc_index |= TC_INDEX_F_SKIP_PROXY;
+	}
+
+	/* Reset packet mark to avoid hitting routing rules again */
+	skb->mark = 0;
+}
+#endif
+
+static inline int rewrite_dmac_if_needed(struct __sk_buff *skb)
+{
+#ifdef FROM_HOST
+	/* When attached to cilium_host, we rewrite the DMAC to the mac of
+	 * cilium_host (peer) to ensure the packet is being considered to be
+	 * addressed to the host (PACKET_HOST) */
+	union macaddr cilium_net_mac = CILIUM_NET_MAC;
+
+	/* Rewrite to destination MAC of cilium_net (remote peer) */
+	if (eth_store_daddr(skb, (__u8 *) &cilium_net_mac.addr, 0) < 0)
+		return send_drop_notify_error(skb, DROP_WRITE_ERROR, TC_ACT_OK);
+#endif
+
+	return TC_ACT_OK;
+}
 
 static inline int handle_ipv6(struct __sk_buff *skb)
 {
 	union v6addr node_ip = { };
-	void *data = (void *) (long) skb->data;
-	void *data_end = (void *) (long) skb->data_end;
-	struct ipv6hdr *ip6 = data + ETH_HLEN;
-	union v6addr *dst = (union v6addr *) &ip6->daddr;
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	union v6addr *dst;
 	int l4_off, l3_off = ETH_HLEN;
 	struct endpoint_info *ep;
 	__u8 nexthdr;
 	__u32 flowlabel;
 	int ret;
 
+	/* If we are attached to cilium_host at egress, this will rewrite the
```

Comment:
why is this above the `(data + l3_off + sizeof(*ip6) > data_end)` check below? Shouldn't we check the packet length first before doing work only to drop it later. I'm guessing the rewrite works because code paths check 'data_end' is past L2 header earlier somewhere.

Reply:
I can move it down a bit but it has to be before `lookup_ip4_endpoint()`. It will void the length checks so we'll have to check it again hence the placement before the initial length check.

Reply:
Fixed

Reply:
great thanks.